File: src/main/java/com/alluvialtrading/sim/VirtualBroker.java
package com.alluvialtrading.sim;


import com.alluvialtrading.algo.Position;
import com.alluvialtrading.data.Quote;
import com.alluvialtrading.data.Trade;
import com.alluvialtrading.tools.TraderLib;
import com.alluvialtrading.vbroker.ASXBackTestingLib;
import com.alluvialtrading.vbroker.BackTestingLib;

import java.io.File;
import java.util.Collections;
import java.util.ArrayList;
import java.util.Date;
import java.util.Hashtable;

/**
 * This class acts as a virtual broker.  It should perform the same functionality as a prime broker.  
 * This means it accepts a list of trades, accepts or rejects based on certain criteria and exits 
 * based on specified criteria.  The input is a CSV file with the columns defined as follows
 * <ol>
 * <li>Market name e.g. ASX</li>
 * <li>Trade date and time e.g. 2010-07-08 16:10:51.797</li>
 * <li>Security name e.g. BHP</li>
 * <li>Trade volume e.g. 4000, -1550. Buy or sell is implied by sign</li>
 * <li>Price e.g. 40.2</li>
 * <li>Exit date/time. If this a number e.g. 60, the broker will interpret this as the number 
 * of seconds from opening. If this is a date such as 2010-04-04 the broker will attempt to exit 
 * on that date.  This must be specified in conjunction with one of the following exit 
 * strategies: OPEN, CLOSE</li>
 * <li>Exit strategy e.g.  Trade.EXIT_AGGRESSIVE, Trade.EXIT_CLOSE, Trade.c, Trade.EXIT_MID,	
 * Trade.EXIT_PASSIVE, Trade.EXIT_VWAP, Trade.EXIT_OPEN</li>
 * </ol>
 * Certain exit strategies require only a date and certain others require both date and time.  
 * The following strategies only require a date: EXIT_CLOSE, EXIT_CLOSE, EXIT_OPEN.  The following
 *  strategies require both date and time: EXIT_MID, EXIT_PASSIVE, EXIT_AGGRESSIVE, EXIT_VWAP.
 * @author dnguyen
 *
 */

public class VirtualBroker {

	// If true manage bankroll limits
	private final boolean MANAGE_BANK = false;

	TraderLib lib = new TraderLib();
	static String importDir = "import";
	ArrayList<Trade> allTrades = new ArrayList<Trade>();
	BackTestingLib connector = new ASXBackTestingLib();
	
	// Trading data
	StringBuffer tradingRecord = new StringBuffer("TradeNo,Security,DateTime,Price,Volume,Brokerage,ReturnOnRisk,Profit,Cash,Shares,Assets\r\n");
	int totalTrades = 0;
	double totalProfit = 0;
	double turnover = 0;
	
	double cash = 100000;
	double shares = 0;
	
	Hashtable<String, Position> openPositions = new Hashtable<String, Position>();
	
	public static void main(String[] args) {
		String[] files = new File(importDir).list();
		for (String file : files) {
			new VirtualBroker(file);
		}
	}
	public VirtualBroker(String tradeData) {

		importTrades(tradeData);		
		lib.writeFile(".", "profit" + tradeData, tradingRecord.toString());
	}
	
	private void doTrade(Trade trade) {

		totalTrades++;

		// TradeNo,Security,Date,Time,Price,Volume,Brokerage,Volatility,ReturnOnRisk,Profit,Cash,Shares,Assets	
		tradingRecord.append(totalTrades + "," + 
				trade.getSecurity() + "," + 
				lib.dateToISODateTimeString(trade.getDate()) + ",");


		double profit = -1;
		
		double price = trade.getPrice();
		int volume = trade.getVolume();
		
		// grab price and volume if necessary
		if (price <= 0) {
			price = connector.getMarketPrice(trade);
		}
		
		tradingRecord.append(price + "," + volume + ",");
		double buyBrokerage = getBrokerage(trade);
		tradingRecord.append(buyBrokerage + ",");
		profit = - buyBrokerage;
		
		Position current = getPosition(trade.getSecurity());
		if (null != current) {
			double currentProfit = current.getPrice() * -volume + price * volume;
			
			// initial sell
			//if (current.getValue() < 0) {
				currentProfit = -currentProfit;
			//}
			profit += currentProfit;			
		}

		totalProfit += profit;
		turnover += Math.abs(price * volume);
		tradingRecord.append(100 * profit / (price * volume) + ",");
		tradingRecord.append(profit);
		
		// store trading position
		storePosition(trade, volume, price);
		
		double currentShares = lib.calculatePosition(openPositions);
		cash = cash + profit;
		double currentCash = cash - currentShares;
		tradingRecord.append("," + currentCash + "," + lib.calculatePosition(openPositions) + 
						"," + (currentCash + currentShares));
		
		if (profit < -30000) {
			//System.out.println("debug");
		}
		tradingRecord.append("\r\n");

	}

	
	private double getBrokerage(Trade trade) {
		
		double value = Math.abs(trade.getPrice() * trade.getVolume());
		if (getBrokerageRate() * value > getBrokerageMin()) {
			return getBrokerageRate() * value;
		}
		else {
			return getBrokerageMin();
		}
		//return 0;
	}
	
	
	public double getBrokerageRate() {
		return 0.001;
	}
	
	public double getBrokerageMin() {
		return 10;
	}
	
	
	/**
	 * Get position in particular stock
	 * @param security
	 * @return
	 */
	private Position getPosition(String security) {
		
		return openPositions.get(security);
		
	}
	
	private void importTrades(String tradeData) {
		
		if (tradeData.charAt(0) == '.') {
			return;
		}
		// Read raw CSV files
		String[] rawData = lib.openFile(importDir, tradeData);	
		int start = 0;
		if (rawData[0].contains("arket")) {
			start = 1;
		}
		// Convert to TradeInfo objects
		for (int i=start; i<rawData.length; i++) {
			
			if (!rawData[i].contains(",")) {
				continue;
			}
			String[] splitLine = rawData[i].split(",");

			String market = splitLine[0];
			Date date = lib.convertISODateTimeString(splitLine[1]);
			String security = splitLine[2];
			int volume = Integer.parseInt(splitLine[3]);
			double price = Double.parseDouble(splitLine[4]);
			String closeWait = splitLine[5];
			String closeStrategy = splitLine[6];
			
			Trade trade = new Trade(market, date, security,
					volume, price, closeWait, closeStrategy);
			allTrades.add(trade);
			
		}
		
		// Go through all the trades and 
		// inject the closing trades when necessary
		int stop = allTrades.size();
		for (int i=0; i<stop ; i++) {
			
			Trade trade = allTrades.get(i);
			// set future time
			if (trade.getCloseWait().length() >= 0) {
				Date newDate = null;
				
				if (trade.getClosingStrategy().equals(Trade.EXIT_VWAP)) {
					
					// calculate VWAP just after
					newDate = new Date(1000 * Integer.parseInt(trade.getCloseWait()) + 
							trade.getDate().getTime());
				}
				else if (trade.getCloseWait().contains("-")) {
					newDate = lib.convertISODateTimeString(trade.getCloseWait());
				}
				else {
					newDate = new Date(1000 * Integer.parseInt(trade.getCloseWait()) + 
							trade.getDate().getTime());
				}
				
				
				Trade close = new Trade(trade.getMarket(), newDate, trade.getSecurity(),
						-trade.getVolume(), 0, "0", trade.getClosingStrategy());
				double exitPrice = 0;
				
				if (trade.getClosingStrategy().equals(Trade.EXIT_VWAP)) {
					exitPrice = connector.getVWAPPrice(close);
				}
				else if (trade.getClosingStrategy().equals(Trade.EXIT_LAST_TRADE)) {
					
					String tradeTime = lib.dateToTimeString(trade.getDate());
					String dateString = lib.dateToISODateString(trade.getDate());
					// If the initial trade occurred after the market closed for 
					// continuous trading, the exit must occur on the subsequent trading day
					if (tradeTime.compareTo(connector.getCloseTime()) > 0) {
						String exitDay = connector.getTradingDate(dateString, 1);
						if (exitDay.contains("-")) {
							exitDay = trade.getCloseWait();
						}
						Trade lastTrade = connector.getLastTrade(trade.getSecurity(), exitDay);
						if (null == lastTrade) {
							trade.setClosingStrategy(Trade.EXIT_AGGRESSIVE);
							Quote lastQuote = connector.getQuote(trade.getSecurity(), exitDay, connector.getCloseTime());
							if (null != lastQuote) {
								exitPrice = connector.getMarketPrice(trade);
								close.setDate(lastQuote.getDateTime());
							}
							else {
								exitPrice = trade.getPrice();
								close.setDate(trade.getDate());
							}
						}
						else {
							exitPrice = lastTrade.getPrice();
							close.setDate(lastTrade.getDate());
						}
						
						
					}
					
					// otherwise exit same day
					else {
						exitPrice = connector.getClosePrice(trade.getSecurity(), lib.convertISODateTimeString(dateString));
					}
					
					
				}
				else {
					exitPrice = connector.getMarketPrice(close);
				}

				
				//System.out.println(exitPrice);
				close.setPrice(exitPrice);
				close.setClosingTrade(true);
				allTrades.add(close);
			}
		}
		
		Collections.sort(allTrades);
		for (Trade trade : allTrades) {
			doTrade(trade);
		}
			
	}
	
	/**
	 * Store the net position in each stock for bankroll management
	 * @param info
	 * @param type
	 */
	private void storePosition(Trade trade, int volume, double price) {	
		
		// store net position in stock
		double stockPosition = volume * price;
		if (volume < 0) {
			stockPosition = -1 * stockPosition;
		}
		
		Position currentPosition = openPositions.get(trade.getSecurity());
		if (null == currentPosition) {
			// new position
			openPositions.put(trade.getSecurity(), new Position(trade, volume, price));
		}
		else {
			currentPosition.addToPosition(volume, price);
			if (0 == currentPosition.getVolume()) {
				openPositions.remove(trade.getSecurity());
			}
		}
		

		
	}
}

File: src/main/scala/com/alluvia/sim/VirtualBroker.scala
package com.alluvia.sim

import com.alluvia.markets.Market
import com.alluvia.database.BackTestingLib

import com.alluvialtrading.algo.Position
import com.alluvialtrading.data.Quote
import com.alluvialtrading.data.Trade
import com.alluvialtrading.tools.TraderLib

import java.util.Date
import java.util.Hashtable

import scala.math.abs

/**
 * This class acts as a virtual broker.  It should perform the same functionality as a prime broker.  
 * This means it accepts a list of trades, accepts or rejects based on certain criteria and exits 
 * based on specified criteria.  The input is a CSV file with the columns defined as follows
 * <ol>
 * <li>Market name e.g. ASX</li>
 * <li>Trade date and time e.g. 2010-07-08 16:10:51.797</li>
 * <li>Security name e.g. BHP</li>
 * <li>Trade volume e.g. 4000, -1550. Buy or sell is implied by sign</li>
 * <li>Price e.g. 40.2</li>
 * <li>Exit date/time. If this a number e.g. 60, the broker will interpret this as the number 
 * of seconds from opening. If this is a date such as 2010-04-04 the broker will attempt to exit 
 * on that date.  This must be specified in conjunction with one of the following exit 
 * strategies: OPEN, CLOSE</li>
 * <li>Exit strategy e.g.  Trade.EXIT_AGGRESSIVE, Trade.EXIT_CLOSE, Trade.c, Trade.EXIT_MID,	
 * Trade.EXIT_PASSIVE, Trade.EXIT_VWAP, Trade.EXIT_OPEN</li>
 * </ol>
 * Certain exit strategies require only a date and certain others require both date and time.  
 * The following strategies only require a date: EXIT_CLOSE, EXIT_CLOSE, EXIT_OPEN.  The following
 *  strategies require both date and time: EXIT_MID, EXIT_PASSIVE, EXIT_AGGRESSIVE, EXIT_VWAP.
 * @author dnguyen
 *
 */

object VirtualBroker {
  def main(args: Array[String]) {
    class MyLSE extends com.alluvia.markets.LSE
    new VirtualBroker("brokertrades.csv", new MyLSE)
  }
}

class VirtualBroker(tradeData: String, market: Market) {
  val importDir: String = "import"
  val connector = new BackTestingLib
  val MANAGE_BANK: Boolean = false
  val lib: TraderLib = new TraderLib

  val tradingRecord: StringBuffer = new StringBuffer("TradeNo,Security,DateTime,Price,Volume,Brokerage,ReturnOnRisk,Profit,Cash,Shares,Assets\r\n")
  val startingCash: Double = 100000 * market.getCurrencyMultiplier
  val openPositions: Hashtable[String, Position] = new Hashtable[String, Position]

  val allTrades = importTrades(tradeData)
  val sortedTrades = allTrades.sortBy(x => x.getDate)
  val record = sortedTrades.scanLeft(new ProfitLine(cash = startingCash)) {
    case (cumulative, trade) =>

      val totalTrades = cumulative.tradeNumber + 1
      val stock = trade.getSecurity
      val date = trade.getDate
      val price = if (trade.getPrice <= 0) connector.getMarketPrice(market.getMarketName, trade) else trade.getPrice
      val volume = trade.getVolume
      val brokerage = getBrokerage(trade)

      val current: Position = getPosition(trade.getSecurity)
      val currentProfit = if (null == trade || null == current) 0 else -(current.getPrice * -volume + price * volume)
      val profit = if (null == trade) -brokerage else -brokerage + currentProfit
      val returnOnRisk = 100 * profit / (price * volume)

      storePosition(trade, volume, price)
      val shares = lib.calculatePosition(openPositions)
      val cash = cumulative.cash + profit
      val currentCash: Double = cash - shares

      val assets = currentCash + shares
      val profitLine = new ProfitLine(totalTrades, stock, date, price, volume, brokerage, returnOnRisk, profit, cash, shares, assets)

      (profitLine)
  }
  for (single <- record.tail) {
    tradingRecord.append(single.toString() + "\r\n")
  }
  lib.writeFile(".", "profit" + tradeData, tradingRecord.toString)

  def getAllTrades(rawData: List[String]): List[Trade] = {

    if (Nil == rawData) return List()
    if (rawData.length == 0) List()
    val line = rawData(0)

    val parse = line.contains(",")
    if (parse) {
      val splitLine: Array[String] = line.split(",")
      val market: String = splitLine(0)
      val date: Date = lib.convertISODateTimeString(splitLine(1))
      val security: String = splitLine(2)
      val volume: Int = Integer.parseInt(splitLine(3))
      val price: Double = (splitLine(4)).toDouble
      val closeWait: String = splitLine(5)
      val closeStrategy: String = splitLine(6)
      val trade: Trade = new Trade(market, date, security, volume, price, closeWait, closeStrategy)
      if (rawData.length == 1) {
        List(trade)
      }
      else {
        trade :: getAllTrades(rawData.tail)
      }
    }
    else
      getAllTrades(rawData.tail)

  }

  private def getBrokerage(trade: Trade): Double = {
    val value: Double = abs(trade.getPrice * trade.getVolume)
    if (getBrokerageRate * value > getBrokerageMin) {
      getBrokerageRate * value
    }
    else {
      getBrokerageMin
    }
  }

  def getBrokerageRate: Double = {
    0.001
  }

  def getBrokerageMin: Double = {
    10
  }

  /**
   * Get position in particular stock
   * @param security
   * @return
   */
  private def getPosition(security: String): Position = {
    openPositions.get(security)
  }

  private def importTrades(tradeData: String): List[Trade] = {
    if (tradeData.charAt(0) == '.') {
      return List()
    }
    val rawData: Array[String] = lib.openFile(importDir, tradeData)
    val start: Int = if (rawData(0).contains("arket")) 1 else 0
    val allTrades = getAllTrades(rawData.toList)

    injectClosingTrades(allTrades)

  }

  def injectClosingTrades(allTrades: List[Trade]): List[Trade] = {

    if (allTrades.length == 0) return List()
    val trade = allTrades.head
    if (trade.getSecurity == "BNKR.L") {
      println("debug")
    }

    val close: Trade = new Trade(trade.getMarket, new Date, trade.getSecurity, -trade.getVolume, 0, "0", trade.getClosingStrategy)

    if (trade.getClosingStrategy == Trade.EXIT_CLOSE) {
         val nextClose = connector.getClosingTrade(market.getMarketName, trade.getSecurity,
           connector.getTradingDate(market.getMarketName, lib.dateToISODateString(trade.getDate), 1), "")
      close.setPrice(nextClose.getPrice)
      close.setClosingTrade(true)
      close.setDate(nextClose.getDate)
           trade :: close :: injectClosingTrades(allTrades.tail)
    }
    else if (trade.getCloseWait.length >= 0) {
      val newDate: Date =
        if (trade.getClosingStrategy == Trade.EXIT_VWAP) {
          new Date(1000 * Integer.parseInt(trade.getCloseWait) + trade.getDate.getTime)
        }
        else if (trade.getCloseWait.contains("-")) {
          lib.convertISODateTimeString(trade.getCloseWait)
        }
        else {
          new Date(1000 * Integer.parseInt(trade.getCloseWait) + trade.getDate.getTime)
        }

      val exitPrice: Double =
        if (trade.getClosingStrategy == Trade.EXIT_VWAP) {
          connector.getVWAPPrice(market.getMarketName, close)
        }
        else if (trade.getClosingStrategy == Trade.EXIT_LAST_TRADE) {
          val tradeTime: String = lib.dateToTimeString(trade.getDate)
          val dateString: String = lib.dateToISODateString(trade.getDate)
          if (tradeTime.compareTo(market.getCloseTime) > 0) {
            val exitDay: String = {
              val result = connector.getTradingDate(market.getMarketName, dateString, 1)
              if (result.contains("-")) trade.getCloseWait else result
            }

            val lastTrade: Trade = connector.getLastTrade(market.getMarketName, trade.getSecurity, exitDay)
            if (null == lastTrade) {
              trade.setClosingStrategy(Trade.EXIT_AGGRESSIVE)
              val lastQuote: Quote = connector.getQuote(market.getMarketName, trade.getSecurity, exitDay, market.getCloseTime)
              if (null != lastQuote) {
                close.setDate(lastQuote.getDateTime)
                connector.getMarketPrice(market.getMarketName, trade)

              }
              else {
                close.setDate(trade.getDate)
                trade.getPrice

              }
            }
            else {
              close.setDate(lastTrade.getDate)
              lastTrade.getPrice
            }
          }
          else {
            connector.getClosePrice(market.getMarketName, trade.getSecurity, lib.convertISODateTimeString(dateString))
          }
        }
        else {
          connector.getMarketPrice(market.getMarketName, close)
        }
      close.setPrice(exitPrice)
      close.setClosingTrade(true)
      trade :: close :: injectClosingTrades(allTrades.tail)
    }
    else injectClosingTrades(allTrades.tail)

  }

  /**
   * Store the net position in each stock for bankroll management
   * @param info
   * @param type
   */
  private def storePosition(trade: Trade, volume: Int, price: Double) {
    val stockPosition: Double = if (volume < 0) -1 * volume * price else volume * price

    val currentPosition: Position = openPositions.get(trade.getSecurity)
    if (null == currentPosition) {
      openPositions.put(trade.getSecurity, new Position(trade, volume, price))
    }
    else {
      currentPosition.addToPosition(volume, price)
      if (0 == currentPosition.getVolume) {
        openPositions.remove(trade.getSecurity)
      }
    }
  }

}

class ProfitLine(
                  val tradeNumber: Int = 0,
                  val security: String = "",
                  val dateTime: Date = new Date,
                  val price: Double = 0,
                  val volume: Double = 0,
                  val brokerage: Double = 0,
                  val returnOnRisk: Double = 0,
                  val profit: Double = 0,
                  val cash: Double = 0,
                  val shares: Double = 0,
                  val assets: Double = 0
                  ) {

  override def toString() = {
    tradeNumber + "," +
      security + "," +
      dateTime + "," +
      price + "," +
      volume + "," +
      brokerage + "," +
      returnOnRisk + "," +
      profit + "," +
      (cash - shares) + "," +
      shares + "," +
      assets
  }
}

